src/js/components/Button/Badge.js

```diff
@@ -0,0 +1,151 @@
+import React, { useContext, useRef, useState } from 'react';
+import styled, { ThemeContext } from 'styled-components';
+import { parseMetricToNum } from '../../utils';
+import { useLayoutEffect } from '../../utils/use-isomorphic-layout-effect';
+
+import { Box } from '../Box';
+import { Stack } from '../Stack';
+import { Text } from '../Text';
+
+const StyledBadgeContainer = styled(Box)`
+  ${props => props.theme.button.badge.container.extend}
+`;
+
+export const Badge = ({ children, content }) => {
+  const theme = useContext(ThemeContext);
+  const contentRef = useRef();
+  const stackRef = useRef();
+
+  const defaultBadgeDimension =
+    typeof content === 'boolean' ||
+    (content && content.value && typeof content.value === 'boolean')
+      ? // empty badge should be smaller. this value was chosen as a default
+        // after experimenting with various values
+        `${parseMetricToNum(theme.button.badge.size.medium) / 2}px`
+      : theme.button.badge.size.medium;
+
+  // size should drive height, match width to height by default
+  // allow width to grow when content is wide
+  const [height, setHeight] = useState(defaultBadgeDimension);
+  const [width, setWidth] = useState(height);
+
+  // scale badge to fit its contents, leaving space horizontally
+  // that is proportional to vertical space
+  useLayoutEffect(() => {
+    const onResize = () => {
+      if (contentRef && contentRef.current) {
+        if (
+          typeof content === 'number' ||
+          (typeof content === 'object' && content.value)
+        ) {
+          const {
+            height: contentHeight,
+            width: contentWidth,
+          } = contentRef.current.getBoundingClientRect();
+
+          // only adjust the width if contentHeight > 0
+          // jest returns 0 for all getBoundingClientRect values,
+          // so this ensures snapshots are closer to correct values
+          if (contentHeight) {
+            // height of content includes extra space around font from
+            // line-height. account for this extra space with 2.5 multiplier
+            // to add proportional horizontal space
+            const verticalSpace =
+              (parseMetricToNum(height) - contentHeight) * 2.5;
+
+            setWidth(
+              `${Math.max(
+                parseMetricToNum(width),
+                Math.ceil(contentWidth + verticalSpace),
+              )}px`,
+            );
+          }
+        } else {
+          // caller has provided custom JSX
+          setWidth(`${contentRef.current.getBoundingClientRect().width}px`);
+          setHeight(`${contentRef.current.getBoundingClientRect().height}px`);
+        }
+      }
+    };
+    window.addEventListener('resize', onResize);
```

Comment:
Why are we tying this to a window resize event? I think we should do this on every layout but only set the state variables if they change.

Reply:
When we resize and hit a responsive breakpoint, the width of the badge can change (because pad is responsive etc.) and so we want to recalculate the width since the offset value is reliant on the dimensions. Otherwise, I was seeing a case where the padding shrunk on small breakpoint but the width wasn't recalculated so the offset of the badge was incorrect.

Reply:
Ah, thanks. Sounds like a good candidate for a comment. :)